File: backend/python/routes/journey.py
# Journey routes for slime navigation and location tracking
# Handles fetching slime current location from journeys table

from fastapi import APIRouter, HTTPException
from python.db import supabase
from python.schemas import JourneyResponse

router = APIRouter(prefix="/journeys", tags=["journeys"])

@router.get("/slimes/{slime_id}/current", response_model=JourneyResponse)
async def get_slime_current_journey(slime_id: str):
    """
    Get the most recent journey for a slime to determine its current location.
    Returns the latest journey entry which contains the slime's current/destination coordinates.
    """
    result = supabase.table("journeys")\
        .select("*")\
        .eq("slime_id", slime_id)\
        .order("created_at", desc=True)\
        .limit(1)\
        .execute()

    if not result.data or len(result.data) == 0:
        raise HTTPException(status_code=404, detail="No journey found for this slime")

    journey = result.data[0]

    return JourneyResponse(
        id=journey["id"],
        slime_id=journey["slime_id"],
        start_lat=journey["start_lat"],
        start_lng=journey["start_lng"],
        dest_lat=journey["dest_lat"],
        dest_lng=journey["dest_lng"],
        progress=journey["progress"],
        status=journey["status"],
        place_name=journey.get("place_name", "Unknown"),
        place_reason=journey.get("place_reason", ""),
        created_at=journey["created_at"]
    )

File: app/lib/journey.ts
// Journey API client for slime location and navigation
// Handles fetching slime current location from journeys table

import { API_BASE_URL } from "./config";

export interface Journey {
  id: string;
  slime_id: string;
  start_lat: number;
  start_lng: number;
  dest_lat: number;
  dest_lng: number;
  progress: number;
  status: string;
  place_name: string;
  place_reason: string;
  created_at: string;
}

/**
 * Get the current/most recent journey for a slime
 * @param slimeId - The slime's ID
 * @returns Journey object or null
 */
export async function getSlimeCurrentLocation(slimeId: string): Promise<Journey | null> {
  try {
    const response = await fetch(`${API_BASE_URL}/journeys/slimes/${slimeId}/current`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.status === 404) {
      return null;
    }

    if (!response.ok) {
      console.warn("Failed to fetch slime location:", await response.text());
      return null;
    }

    const journey: Journey = await response.json();
    return journey;
  } catch (error) {
    console.warn("Error fetching slime location:", error);
    return null;
  }
}

/**
 * Calculate distance between two coordinates using Haversine formula
 * @returns Distance in meters
 */
export function calculateDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371000; // Earth's radius in meters
  const φ1 = (lat1 * Math.PI) / 180;
  const φ2 = (lat2 * Math.PI) / 180;
  const Δφ = ((lat2 - lat1) * Math.PI) / 180;
  const Δλ = ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
    Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c; // Distance in meters
}

/**
 * Check if user is within acceptable range of slime location
 * @param userLat - User's latitude
 * @param userLng - User's longitude
 * @param slimeLat - Slime's latitude
 * @param slimeLng - Slime's longitude
 * @param maxDistanceMeters - Maximum allowed distance (default 5000m / 5km)
 * @returns true if user is close enough
 */
export function isNearSlime(
  userLat: number,
  userLng: number,
  slimeLat: number,
  slimeLng: number,
  maxDistanceMeters: number = 5000
): boolean {
  const distance = calculateDistance(userLat, userLng, slimeLat, slimeLng);
  console.log(`Distance to slime: ${distance.toFixed(2)}m (${(distance / 1000).toFixed(2)}km)`);
  return distance <= maxDistanceMeters;
}
